### ISSUE TITLE
"Cycle detected in the references of '/nix/store/...'" message could be more useful

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#481]
It doesn't seem like the sort of thing I can easily debug with my usual `nix-build -K`, so I'm kind of stumped as to where the cycle comes from. If it could print some representation of the cycle, that'd be quite a bit more helpful!

**Comment by shlevy** [comment]
This isn't ideal, but after the build the output paths should still be in the store and you can grep for the hashes of other outputs in each

(this error only results from a multiple outputs build of course)

**Comment by copumpkin** [comment]
Why is it even an error, actually?

**Comment by shlevy** [comment]
Ideally it wouldn't be, just no one has made sure nothing enters an infinite loop or anything if there's a cycle in the dependency graph so this is safest.

**Comment by vcunat** [comment]
The acyclicity of nix store dependencies is a useful property. It allows registering validity of paths one by one, collecting garbage by reference counting, etc. Note that cycles always have to be together anyway (strongly-connected components), so we may just as well collapse them and have each strong-component within a single output.

**Comment by copumpkin** [comment]
Well, a given output can already refer to itself, right? So the GC must
already have some code (perhaps special-cased) to deal with that.

Allowing cycles in multiple outputs would greatly simplify splitting up a
package I have, since currently I have to compute its SCCs myself by hand
and it seems like something the machine should do for me :)

**Comment by shlevy** [comment]
I disagree with the "may as well collapse" bit, FWIW. Having the files exist in different prefixes can be useful even if they have to exist in the store as a unit. For example, consider nix-env: it will only treat the specific path as something to be installed, even though it will require the outputs it depends on to still be in the store.

**Comment by edolstra** [comment]
Yes, self-references are special-cased.

And I agree that sometimes it's useful to allow cycles, especially for generic build functions where you don't know in advance which outputs will reference each other. E.g. in `buildPerlPackage` I might want to have `bin` and `lib` outputs, but for some Perl packages those might be cyclical.

**Comment by copumpkin** [comment]
Yeah, the lack of up-front knowledge is basically my situation. Should I
just rename this ticket to "allow cyclic references in multi-output
derivations"?

**Comment by edolstra** [comment]
Better make a separate issue, the error message is more likely to be fixed soon :smiley:

**Comment by vcunat** [comment]
Cycles would also make quite some interaction with intensional names for the cyclic outputs #296. It would be solvable, I expect, only complicating it nontrivially.

**Comment by datakurre** [comment]
So, how this could be debugged? I didn't fully understand @shlevy's tip. I'm seeing this while trying to build mariadb on darwin (it has multiple outputs = ["out", "lib"] and lib gives the cyclic error).

**Comment by vcunat** [comment]
I do `$ grep hash-of-one-output -R /nix/store/the-other-output` to see the references.

**Comment by datakurre** [comment]
@vcunat Thanks.

**Comment by domenkozar** [comment]
see https://github.com/NixOS/nix/pull/925

**Comment by domenkozar** [comment]
Could the error use `why-depends` command to show the underlying cause? I'm sometimes discouraged in digging into multiple outputs issues because of clumsiness to debug them :)

**Comment by jtojnar** [comment]
This is still an issue. Domen’s suggestion would be very helpful, especially when building on CI for a platform you do not have access to to debug.

**Comment by Artturin** [comment]
https://github.com/NixOS/nix/pull/6585

**Comment by arianvp** [comment]
> I do $ grep hash-of-one-output -R /nix/store/the-other-output to see the references.

I don't understand this hint at all.  Could you elaborate? The only thing the error prints out is the `.drv` file. 

And the outputs don't exist in the nix store as the built failed due to the cycle.

**Comment by charles-dyfis-net** [comment]
> And the outputs don't exist in the nix store as the built failed due to the cycle.

Despite the failure, the outputs _are_ left in the store.

**Comment by peterwaller-arm** [comment]
> Despite the failure, the outputs are left in the store.

If this is so, it would be great if the error message shown by nix could include that fact and show the paths; build errors usually imply the built artefact is not present, so it would be a reasonable assumption to make.
